src/charm.py

```diff
@@ -759,19 +870,52 @@ def _reconcile_mongo_hosts_and_users(self, event: RelationEvent) -> None:
                 logger.info("Deferring reconfigure: error=%r", e)
                 event.defer()
 
+    def __handle_partition_on_stop(self) -> None:
+        """Raise partition to prevent other units from restarting if an upgrade is in progress.
+
+        If an upgrade is not in progress, the leader unit will reset the partition to 0.
+        """
+        current_unit_number = unit_number(self.unit)
+        if k8s_upgrade.partition.get(app_name=self.app.name) < current_unit_number:
+            k8s_upgrade.partition.set(app_name=self.app.name, value=current_unit_number)
+            logger.debug(f"Partition set to {current_unit_number} during stop event")
+
+    def __handle_relation_departed_on_stop(self):
+        """Leave replicaset."""
+        logger.debug(f"{self.unit.name} blocking on_stop")
+        is_in_replica_set = True
+        timeout = UNIT_REMOVAL_TIMEOUT
+        while is_in_replica_set and timeout > 0:
+            is_in_replica_set = self.is_unit_in_replica_set()
+            time.sleep(1)
+            timeout -= 1
+            if timeout < 0:
+                raise Exception(f"{self.unit.name}.on_stop timeout exceeded")
+        logger.debug("{self.unit.name} releasing on_stop")
+        self.unit_departed = False
+
+    def __handle_upgrade_on_stop(self):
+        """Sets the unit state to RESTARTING and step down from replicaset."""
```

Comment:
similar warning as the docstring above - but note that we could have downtime if the primary is not re-elected in time

Reply:
3fe85961